### ISSUE TITLE
modern D array indexing/slicing

### ISSUE THREAD

**Issue by John-Colvin** [kyllingstad/scid#24]
A couple of notes:

MatrixView has opIndex return by ref in order to allow ```array[a..b] op= v;```. The should be done with opIndexOpAssign instead.

With the new extended semantics of opIndex and opSlice, a proper NDArray type can be implemented.

http://dlang.org/operatoroverloading.html

**Comment by kyllingstad** [comment]
> MatrixView has opIndex return by ref in order to allow `array[i,j] op= v`;. The should be done with opIndexOpAssign instead.

Good point! `opIndexOpAssign` didn't exist when this code was written. That would take care of the issue with the zero elements in a triangular matrix being modifiable. I'll fix it.

> With the new extended semantics of opIndex and opSlice, a proper NDArray type can be implemented, with multi-dimensional slicing.

I know, and agree that would be nice to have. However, if I were to dedicate some time to doing something with the new slicing features, I think I would prioritise implementing submatrix slicing first, before going all out and creating a full-fledged N-dimensional array type. (Also, that seems general enough to be in Phobos, if you ask me.)

**Comment by 9il** [comment]
You can copy-past if it is needed
https://github.com/9il/simple_matrix/blob/master/source/simple_matrix.d#L432

**Comment by 9il** [comment]
Code example from corresponding [discussion](http://forum.dlang.org/thread/eubhesirsvtnqowbevdm@forum.dlang.org?page=4):

```D
size_t anyNumber;
auto ar = new int[3 * 8 * 9 + anyNumber];
auto slice = Slice[0..3, 4..8, 1..9];
assert(ar.canBeSlicedWith(slice)); //checks that ar.length <= 3 * 
8 * 9

auto tensor = ar.sliced(slice);
tensor[0, 1, 2] = 4;

auto matrix = tensor[0..$, 1, 0..$];
assert(matrix[0, 2] == 4);
```

**Comment by 9il** [comment]
`Slice` implementation is very simple:
```D
struct Slice
{
static @safe pure nothrow @nogc:

	size_t[2] opSlice(size_t p)(size_t a, size_t b)
	in {
		assert(a <= b);
	}
	body {
		return [a, b];
	}
	
	auto opIndex(Args...)(Args args)
	{
		import std.typecons : tuple;
		return tuple(args);
	}
}

///
unittest {
	import std.typecons : tuple;
	auto slice = Slice[0..1, 4, 2..3];
	assert(slice == tuple([0, 1], 4, [2, 3]));
}
```

**Comment by 9il** [comment]
Phobos PR https://github.com/D-Programming-Language/phobos/pull/3397

**Comment by kyllingstad** [comment]
Nice! If this goes through in Phobos, I guess `MatrixView` can be removed from SciD altogether? Your `Slice` seems to do the same job.

**Comment by 9il** [comment]
Yes. Current (example in Phobos PR) concept is simplified, but logic is the same.
